ibis/common/tests/test_grounds.py

```diff
@@ -0,0 +1,495 @@
+        __slots__ = ('_memoize',)  # new slot
+        sep = ValidatorFunction(str)  # new overridden slot
+
+    assert Op.__slots__ == ('_cache', 'arg')
+    assert StringOp.__slots__ == ('arg',)
+    assert StringSplit.__slots__ == ('sep',)
+    assert StringJoin.__slots__ == ('_memoize', 'sep')
+
+
+def test_multiple_inheritance():
+    # multiple inheritance is allowed only if one of the parents has non-empty
+    # __slots__ definition, otherwise python will raise lay-out conflict
+
+    class Op(Annotable):
+        __slots__ = ('_hash',)
+
+    class ValueOp(Annotable):
+        arg = InstanceOf(object)
+
+    class Reduction(ValueOp):
+        _reduction = True
+
+    class UDF(ValueOp):
+        func = ValidatorFunction(lambda fn, this: fn)
+
+    class UDAF(UDF, Reduction):
+        arity = IsInt
+
+    class A(Annotable):
+        a = IsInt
+
+    class B(Annotable):
+        b = IsInt
+
+    msg = "multiple bases have instance lay-out conflict"
+    with pytest.raises(TypeError, match=msg):
+
+        class AB(A, B):
+            ab = IsInt
+
+    assert UDAF.__slots__ == ('arity',)
+    strlen = UDAF(arg=2, func=lambda value: len(str(value)), arity=1)
+    assert strlen.arg == 2
+    assert strlen.arity == 1
+    assert strlen._reduction is True
+
+
+@pytest.mark.parametrize(
+    "obj",
+    [
+        StringOp(arg="something"),
+        Parameter("test"),
+    ],
+)
+def test_pickling_support(obj):
+    assert_pickle_roundtrip(obj)
+
+
+def test_multiple_inheritance_argument_order():
+    class ValueOp(Annotable):
+        arg = IsAny
+
+    class VersionedOp(ValueOp):
+        version = IsInt
+
+    class Reduction(Annotable):
+        _reduction = True
+
+    class Sum(VersionedOp, Reduction):
+        where = Optional(IsBool, default=False)
+
+    assert Sum._reduction is True
+    assert str(Sum.__signature__) == "(arg, version, where=None)"
+
+
+def test_multiple_inheritance_optional_argument_order():
+    class ValueOp(Annotable):
+        pass
+
+    class ConditionalOp(Annotable):
+        where = Optional(IsBool, default=False)
+
+    class Between(ValueOp, ConditionalOp):
+        min = IsInt
+        max = IsInt
+        how = Optional(IsStr, default="strict")
+
+    assert str(Between.__signature__) == "(min, max, how=None, where=None)"
+
+
+def test_immutability():
+    class ValueOp(Annotable):
+        a = IsInt
+
+    op = ValueOp(1)
+    with pytest.raises(TypeError):
+        op.a = 3
+
+
+class Node(Comparable):
+
+    # override the default cache object
+    __cache__ = WeakCache()
+    num_equal_calls = 0
+
+    def __init__(self, name):
+        self.name = name
+
+    def __hash__(self):
+        return hash(self.name)
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        return f"Node(name={self.name})"
+
+    def __equals__(self, other):
+        Node.num_equal_calls += 1
+        return self.name == other.name
+
+
+@pytest.fixture
+def cache():
```

Comment:
Can you move the cache tests into `tests/test_caching.py`? Ideally we closely match the module structure of the parent.

Reply:
This tests `Comparable` which is implemented in `grounds.py`

Reply:
Ah, ok. Fine as is then!